runtime/compact.c

```diff
@@ -275,75 +221,58 @@ static void do_compaction (intnat new_allocation_policy)
 
       chend = ch + Chunk_size (ch);
       while ((char *) p < chend){
-        word q = *p;
+        header_t h = * (header_t *) p;
+        size_t sz;
 
-        if (Ecolor (q) == 0 || Tag_ehd (q) == Infix_tag){
-          /* There were (normal or infix) pointers to this block. */
-          size_t sz;
+        while (Is_gray_hd (h)) h = * dptr (h);
+        sz = Whsize_hd (h);
+
+        CAMLassert (!Is_black_hd (h));
+        CAMLassert (!Is_gray_hd (h));
+        if (Is_white_hd (h)){
+          word q;
           tag_t t;
           char *newadr;
-#ifdef WITH_PROFINFO
-          uintnat profinfo;
-#endif
-          word *infixes = NULL;
 
-          while (Ecolor (q) == 0) q = * (word *) q;
-          sz = Whsize_ehd (q);
-          t = Tag_ehd (q);
-#ifdef WITH_PROFINFO
-          profinfo = Profinfo_ehd (q);
-#endif
-          if (t == Infix_tag){
-            /* Get the original header of this block. */
-            infixes = p + sz;
-            q = *infixes;
-            CAMLassert (Ecolor (q) == 2);
-            while (Ecolor (q) != 3) q = * (word *) (q & ~(uintnat)3);
-            sz = Whsize_ehd (q);
-            t = Tag_ehd (q);
-          }
+          t = Tag_hd (h);
+          CAMLassert (t != Infix_tag);
 
           newadr = compact_allocate (Bsize_wsize (sz));
           q = *p;
-          while (Ecolor (q) == 0){
-            word next = * (word *) q;
-            * (word *) q = (word) Val_hp (newadr);
-            q = next;
+          while (Is_gray_hd (q)){
+            word *pp = dptr (q);
```

Comment:
The house style is `word * pp`.

Reply:
It's not idiomatic but if you insist I won't fight you over this.

Reply:
Forget about this.  I still prefer the "spaces on both sides" style, but let's not get distracted.